### PR TITLE
WIP: Work on getting CI targets working with the build-tools container

### PR DESCRIPTION
**Pull request by geeknoid**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -29,7 +29,7 @@
 export BUILD_WITH_CONTAINER ?= 0
 
 ifeq ($(BUILD_WITH_CONTAINER),1)
-IMG = gcr.io/istio-testing/build-tools:2019-08-25T20-37-02
+IMG = gcr.io/istio-testing/build-tools:2019-08-28T11-32-19
```
Comment (rlenglet) [inline Makefile]: This image doesn't exist.
```suggestion
IMG = gcr.io/istio-testing/build-tools:2019-08-29T13-57-48
```

**Comment by rlenglet** [comment]
I can successfully run `make docker test-cni` from macos with the following patch:
```
diff --git a/Makefile b/Makefile
index 94e9d57e..93a604ad 100644
--- a/Makefile
+++ b/Makefile
@@ -29,7 +29,7 @@
 export BUILD_WITH_CONTAINER ?= 0
 
 ifeq ($(BUILD_WITH_CONTAINER),1)
-IMG = gcr.io/istio-testing/build-tools:2019-08-28T11-32-19
+IMG = gcr.io/istio-testing/build-tools:2019-08-29T13-57-48
 UID = $(shell id -u)
 PWD = $(shell pwd)
 GOBIN_SOURCE ?= $(GOPATH)/bin
diff --git a/Makefile.core.mk b/Makefile.core.mk
index f8715e9e..dcce87a2 100644
--- a/Makefile.core.mk
+++ b/Makefile.core.mk
@@ -264,9 +264,9 @@ else
 endif
 test-cni: | $(JUNIT_REPORT)
        mkdir -p $(dir $(JUNIT_UNIT_TEST_XML))
-       set -o pipefail; \
+       set -o pipefail; unset GOOS && unset GOARCH && CGO_ENABLED=1 && \
        $(MAKE) -f Makefile.core.mk --keep-going $(TEST_OBJ) \
-       2>&1 | tee >($(JUNIT_REPORT) > $(JUNIT_UNIT_TEST_XML))
+               2>&1 | tee >($(JUNIT_REPORT) > $(JUNIT_UNIT_TEST_XML))
 
 .PHONY: install-test
 # May want to make this depend on push but it will always push at the moment:  install-test: docker.push
diff --git a/tools/istio-cni-docker.mk b/tools/istio-cni-docker.mk
index f3e0bf09..b9b5d414 100644
--- a/tools/istio-cni-docker.mk
+++ b/tools/istio-cni-docker.mk
@@ -43,8 +43,9 @@ docker.install-cni: $(ISTIO_OUT)/istio-cni tools/packaging/common/istio-iptables
                deployments/kubernetes/install/scripts/filter.jq
        mkdir -p $(ISTIO_DOCKER)/install-cni
        cp $^ $(ISTIO_DOCKER)/install-cni
-       time (cd $(ISTIO_DOCKER)/install-cni && \
-               docker build -t $(HUB)/install-cni:$(TAG) -f Dockerfile.install-cni .)
+       time docker build -t $(HUB)/install-cni:$(TAG) \
+               -f $(ISTIO_DOCKER)/install-cni/Dockerfile.install-cni \
+               $(ISTIO_DOCKER)/install-cni
 
 DOCKER_TARGETS:=docker.install-cni
 
```

Trying `make prow-e2e` now.

**Comment by rlenglet** [comment]
Hmmm actually the tests still fail in `make test-cni`.

**Comment by rlenglet** [comment]
I almost got it to work.
Now I'm blocked by the fact that the `install_cni.go` test is itself running Docker containers, into which it mounts 3 temporary subdirectories from its own `/tmp`, which of course don't exist in Docker's host (they exist only in the container running `install_cni.go`), so those mounts are empty.
I'll now try mounting `/tmp' from the host into the test container's `/tmp`...

**Comment by rlenglet** [comment]
OK, this patch makes `make docker` and `make test-cni` run for real:
```
diff --git a/Makefile b/Makefile
index 94e9d57e..7280406f 100644
--- a/Makefile
+++ b/Makefile
@@ -29,7 +29,7 @@
 export BUILD_WITH_CONTAINER ?= 0
 
 ifeq ($(BUILD_WITH_CONTAINER),1)
-IMG = gcr.io/istio-testing/build-tools:2019-08-28T11-32-19
+IMG = gcr.io/istio-testing/build-tools:2019-08-29T13-57-48
 UID = $(shell id -u)
 PWD = $(shell pwd)
 GOBIN_SOURCE ?= $(GOPATH)/bin
@@ -67,11 +67,13 @@ RUN = docker run -t -i --sig-proxy=true -u $(UID) --rm \
        -v /etc/passwd:/etc/passwd:ro \
        -v $(readlink /etc/localtime):/etc/localtime:ro \
        -v /var/run/docker.sock:/var/run/docker.sock \
+       -v /tmp:/tmp \
        $(CONTAINER_OPTIONS) \
        --mount type=bind,source="$(PWD)",destination="/work" \
        --mount type=volume,source=istio-go-mod,destination="/go/pkg/mod" \
        --mount type=volume,source=istio-go-cache,destination="/gocache" \
        --mount type=bind,source="$(GOBIN_SOURCE)",destination="/go/out/bin" \
+       --mount type=tmpfs,destination=/run/secrets/kubernetes.io/serviceaccount \
        -w /work $(IMG)
 else
 export GOBIN ?= ./out/bin
diff --git a/Makefile.core.mk b/Makefile.core.mk
index f8715e9e..77c28133 100644
--- a/Makefile.core.mk
+++ b/Makefile.core.mk
@@ -262,9 +262,12 @@ ifeq ($(WHAT),)
 else
        TEST_OBJ = selected-pkg-test
 endif
-test-cni: | $(JUNIT_REPORT)
+/var/run/secrets/kubernetes.io/serviceaccount/token:
+       echo "dummy-token" > /var/run/secrets/kubernetes.io/serviceaccount/token
+
+test-cni: /var/run/secrets/kubernetes.io/serviceaccount/token | $(JUNIT_REPORT)
        mkdir -p $(dir $(JUNIT_UNIT_TEST_XML))
-       set -o pipefail; \
+       set -o pipefail; unset GOOS && unset GOARCH && CGO_ENABLED=1 && \
        $(MAKE) -f Makefile.core.mk --keep-going $(TEST_OBJ) \
        2>&1 | tee >($(JUNIT_REPORT) > $(JUNIT_UNIT_TEST_XML))
 
diff --git a/tools/istio-cni-docker.mk b/tools/istio-cni-docker.mk
index f3e0bf09..b9b5d414 100644
--- a/tools/istio-cni-docker.mk
+++ b/tools/istio-cni-docker.mk
@@ -43,8 +43,9 @@ docker.install-cni: $(ISTIO_OUT)/istio-cni tools/packaging/common/istio-iptables
                deployments/kubernetes/install/scripts/filter.jq
        mkdir -p $(ISTIO_DOCKER)/install-cni
        cp $^ $(ISTIO_DOCKER)/install-cni
-       time (cd $(ISTIO_DOCKER)/install-cni && \
-               docker build -t $(HUB)/install-cni:$(TAG) -f Dockerfile.install-cni .)
+       time docker build -t $(HUB)/install-cni:$(TAG) \
+               -f $(ISTIO_DOCKER)/install-cni/Dockerfile.install-cni \
+               $(ISTIO_DOCKER)/install-cni
 
 DOCKER_TARGETS:=docker.install-cni
 
```

**Comment by rlenglet** [comment]
And `make prow-e2e` cannot run because the build image is missing `kind`.

**Comment by tiswanso** [comment]
> And `make prow-e2e` cannot run because the build image is missing `kind`.

Could we use the same image as prow presubmits to get KinD (and build)?  (admittedly, I'm way out-of-date from when I really understood the details of istio/test-infra/prow)
https://github.com/istio/test-infra/blob/master/prow/cluster/jobs/istio/cni/istio.cni.master.yaml#L19

**Comment by geeknoid** [comment]
@rlenglet Wow, thanks Romain for making such progress.

I'll update my PR with your changes, and I'll get Kind in the build-tools image so we can go to the next steps.

**Comment by geeknoid** [comment]
@tiswanso We're trying to converge all repos to be building with the same build-tools container image. We've been gradually expanding what is in the container as we bring up the different repos. So hitting a wall where a tool is needed in the container is normal. I'll go add what is needed in the container shortly and the iterative loop to get this repo converted to the container can resume...

**Comment by tiswanso** [comment]
> @rlenglet Wow, thanks Romain for making such progress.

+1 ... great progress

**Comment by rlenglet** [comment]
Note that the tests will write some (small) temp files into the host's `/tmp`, and that can't be fixed without being more invasive, e.g. mouting `tmpfs` instead of `/tmp` in the nested containers.

**Comment by tiswanso** [comment]
> Note that the tests will write some (small) temp files into the host's `/tmp`, and that can't be fixed without being more invasive, e.g. mouting `tmpfs` instead of `/tmp` in the nested containers.

Right... that's existing behavior so I don't think it needs to be changed in this PR.  IIRC, we originally did that so we could debug failures.

**Comment by geeknoid** [comment]
@rlenglet I've updated my PR with the latest common files, which brings in the latest build-tools image that includes KinD. I tried applying your patch, but git complains that it is corrupt. So could you try again, given this updated PR please?

**Comment by sdake** [comment]
from the check:
```
+ pushd ../istio
/home/prow/go/src/istio.io/istio /home/prow/go/src/istio.io/cni
+ make istioctl
STATIC=0 GOOS=linux GOARCH=amd64 LDFLAGS='-extldflags -static -s -w' bin/gobuild.sh /home/prow/go/out/linux_amd64/release/istioctl ./istioctl/cmd/istioctl
flag provided but not defined: -trimpath
usage: go build [-o output] [-i] [build flags] [packages]
Run 'go help build' for details.
Makefile.core.mk:393: recipe for target 'istioctl' failed
```

**Comment by sdake** [comment]
```
Stevens-MacBook-Pro:pilot sdake$ go build -trimpath
flag provided but not defined: -trimpath
usage: go build [-o output] [-i] [build flags] [packages]
Stevens-MacBook-Pro:pilot sdake$ go version
go version go1.12.7 darwin/amd64
```

The internet says -trimpath was added in golang 1.13.

**Comment by sdake** [comment]
From buildmodern check:
```
time docker build -t istio/install-cni:fc050a1a908b229997ed303fa5de5187c6e549c4 \
	-f /home/prow/go/out/linux_amd64/release/docker_temp/install-cni/Dockerfile.install-cni \
	/home/prow/go/out/linux_amd64/release/docker_temp/install-cni
time="2019-09-19T18:46:00Z" level=error msg="failed to dial gRPC: cannot connect to the Docker daemon. Is 'docker daemon' running on this host?: dial unix /var/run/docker.sock: connect: no such file or directory"
context canceled```
this implies Kind does not expose /var/run/docker.sock of the host K8s system.  This is not unexpected.  Using /var/run/docker.sock iin K8s from docker is dangerous.  Especially so with K8s in docker in docker.

**Comment by sdake** [comment]
Looking at the buildtools container, it clearly indicates 1.13.0 of golang is in use.

**Comment by sdake** [comment]
go version is 1.12.9 because istio-builder is being used.

```
/home/prow/go/src/istio.io/istio /home/prow/go/src/istio.io/cni
+ make istioctl
STATIC=0 GOOS=linux GOARCH=amd64 LDFLAGS='-extldflags -static -s -w' bin/gobuild.sh /home/prow/go/out/linux_amd64/release/istioctl ./istioctl/cmd/istioctl
go version go1.12.9 linux/amd64
```

See PR: https://github.com/istio/test-infra/pull/1829 which updates istio-builder.

Note istio-builder is the wrong container entirely for this purpose - the goal is to be using buildtools for a consistent toolchain.

**Comment by sdake** [comment]
@tiswanso I assume the circleci job is desireable because ipv6 doesn't work in kind?

Could you verify?  If the job is redundant, we are trying to limit our use of circleci and I'll remove it from the list of jobs.  PLMK.

Cheers
-steve

**Comment by sdake** [comment]
buildmodern_cni is using the build-tools container (which contains no entrypoint to launch docker: https://github.com/istio/test-infra/blob/master/prow/cluster/jobs/istio/cni/istio.cni.release-1.3.gen.yaml#L279-L288

**Comment by sdake** [comment]
e2e (which was previously failing in this PR) is using istiobulider with an entrypoint to start docker: https://github.com/istio/test-infra/blob/master/prow/cluster/jobs/istio/cni/istio.cni.release-1.3.gen.yaml#L235-L244

The conclusion is we need to subsume the istiobuilder docker entrypoint work into build-tools.  THis may take multiple revisions to test-infra and tools/docker/build-tools to obtain a working check that functions well on prow.

**Comment by howardjohn** [comment]
I was unable to get ipv6 working on kind, did to the underlying prow nodes not supporting ipv6. I agree you need the entry point but it will require changes to the docker files since we blow away some of the files required to `service docker start`

**Comment by sdake** [comment]
@howardjohn I understand.  I will be taking on making docker run within the build-tools container today and hopefully finish.  At latest, my target is this week.  As for the dind-ipv6 job, it was already failing prior to this pr.  See #180, #181, or #182.

Cheers
-steve

**Comment by howardjohn** [comment]
Sounds great, thanks!!

**Pull request closed without merging**